textpipe/wrappers.py

```diff
@@ -0,0 +1,107 @@
+"""
+Wrappers around classes of external libraries used in Doc class.
+"""
+
+import bz2
+import numpy as np
+import pickle
+from tqdm import tqdm
+from functools import lru_cache
+
+from gensim.models.keyedvectors import KeyedVectors
+from redis import Redis
+from redis.exceptions import RedisError
+
+
+class RedisKeyedVectorException(Exception):
+    """ Raised when something while loading"""
+    pass
+
+
+class RedisKeyedVectors(KeyedVectors):
+    """
+    Class to imitate gensim's KeyedVectors, but instead getting the vectors from the memory,
+    the vectors will be retrieved from a redis db.
+    Based on https://engineering.talentpair.com/serving-word-vectors-for-distributed-computations-c5065cbaa02f
+    """
+
+    def __init__(self, host, port, db=0, key='', max_lru_cache_size=1024):
+        self.word_vec = lru_cache(maxsize=max_lru_cache_size)(self.word_vec)
+        self.syn0 = []
+        self.syn0norm = None
+        self.index2word = []
+        self.key = f'w2v_{key}'
+
+        try:
+            self._redis = Redis(host, port, db)
+        except RedisError as e:
+            raise RedisKeyedVectorException(f'The connection to Redis failed while trying to '
+                                            f'initiate the client. Redis error message: {e}')
+
+    @classmethod
+    def load_word2vec_format(cls, **kwargs):
+        raise NotImplementedError(
+            "You can't load a word model that way. It needs to pre-loaded into redis")
+
+    def save(self, *args, **kwargs):
+        raise NotImplementedError("You can't write back to Redis that way.")
+
+    def save_word2vec_format(self, **kwargs):
+        raise NotImplementedError("You can't write back to Redis that way.")
+
+    def word_vec(self, word):
+        """
+        This method is mimicking the word_vec method from the Gensim KeyedVector class. Instead of
+        looking it up from an in memory dict, it
+        - requests the value from the redis instance, where the key is a combination between
+        an optional word vector model key and the word itself
+        - decompresses it
+        - and finally unpickles it
+
+        :param word: string
+
+        :returns: numpy array of dim of the word vector model (for Google: 300, 1)
+        """
+
+        try:
+            cache_entry = self._redis.hget(self.key, word)
+            if not cache_entry:
+                raise KeyError(f'Key {cache_entry} does not exist in cache')
+            return pickle.loads(bz2.decompress(cache_entry))
```

Comment:
bz2 compression needed by the library? Else snappy is probably faster and uses less CPU.

Reply:
Good point, using snappy now!

Reply:
Reverted to bz2 again, turns out snappy introduces a dependency to the system (you have to install the snappy C lib on you system).

Reply:
IIRC bzip2 is one of the slowest compression algos out there. Are these floating point vectors that are being compressed? What is the expected compression ratio?

Reply:
Good point:
```
np.mean([getsizeof(pickle.dumps(np.random.rand(300))) / getsizeof(bz2.compress(pickle.dumps(np.random.rand(300)))) for _ in range(1000)])

0.8810087930763757
```
...with an average compression ratio of 0.88 I'd say it's not worth it. I'll remove the compression.